cmd/nerdctl/cp_linux.go

```diff
@@ -0,0 +1,270 @@
+}
+
+// kopy implements `nerdctl cp`.
+//
+// See https://docs.docker.com/engine/reference/commandline/cp/ for the specification.
+func kopy(ctx context.Context, container2host bool, pid int, dst, src string, followSymlink bool) error {
+	tarBinary, isGNUTar, err := tarutil.FindTarBinary()
+	if err != nil {
+		return err
+	}
+	var srcFull, dstFull string
+	root := fmt.Sprintf("/proc/%d/root", pid)
+	if container2host {
+		srcFull, err = securejoin.SecureJoin(root, src)
+		dstFull = dst
+	} else {
+		srcFull = src
+		dstFull, err = securejoin.SecureJoin(root, dst)
+	}
+	if err != nil {
+		return err
+	}
+	var (
+		srcIsDir       bool
+		dstExists      bool
+		dstExistsAsDir bool
+	)
+	if st, err := os.Stat(srcFull); err != nil {
+		return err
+	} else {
+		srcIsDir = st.IsDir()
+	}
+	// dst may not exist yet, so err is negligible
+	if st, err := os.Stat(dstFull); err == nil {
+		dstExists = true
+		dstExistsAsDir = st.IsDir()
+	}
+	dstEndsWithSep := strings.HasSuffix(dst, string(os.PathSeparator))
+	srcEndsWithSlashDot := strings.HasSuffix(src, string(os.PathSeparator)+".")
+	if !srcIsDir && dstEndsWithSep && !dstExistsAsDir {
+		// The error is specified in https://docs.docker.com/engine/reference/commandline/cp/
+		// See the `DEST_PATH does not exist and ends with /` case.
+		return fmt.Errorf("the destination directory must exists: %w", err)
+	}
+	if !srcIsDir && srcEndsWithSlashDot {
+		return fmt.Errorf("the source is not a directory")
+	}
+	if srcIsDir && dstExists && !dstExistsAsDir {
+		return fmt.Errorf("cannot copy a directory to a file")
+	}
+	if srcIsDir && !dstExists {
+		if err := os.MkdirAll(dstFull, 0755); err != nil {
+			return err
+		}
+	}
+
+	var tarCDir, tarCArg string
+	if srcIsDir {
+		if !dstExists || srcEndsWithSlashDot {
+			// the content of the source directory is copied into this directory
+			tarCDir = srcFull
+			tarCArg = "."
+		} else {
+			// the source directory is copied into this directory
+			tarCDir = filepath.Dir(srcFull)
+			tarCArg = filepath.Base(srcFull)
+		}
+	} else {
+		// Prepare a single-file directory to create an archive of the source file
+		td, err := os.MkdirTemp("", "nerdctl-cp")
+		if err != nil {
+			return err
+		}
+		defer os.RemoveAll(td)
+		tarCDir = td
+		cp := []string{"cp", "-a"}
+		if followSymlink {
+			cp = append(cp, "-L")
+		}
+		if dstEndsWithSep || dstExistsAsDir {
+			tarCArg = filepath.Base(srcFull)
+		} else {
+			// Handle `nerdctl cp /path/to/file some-container:/path/to/file-with-another-name`
+			tarCArg = filepath.Base(dstFull)
+		}
+		cp = append(cp, srcFull, filepath.Join(td, tarCArg))
+		cpCmd := exec.CommandContext(ctx, cp[0], cp[1:]...)
+		logrus.Debugf("executing %v", cpCmd.Args)
+		if out, err := cpCmd.CombinedOutput(); err != nil {
+			return fmt.Errorf("failed to execute %v: %w (out=%q)", cpCmd.Args, err, string(out))
+		}
+	}
+	tarC := []string{tarBinary, "-c", "-f"}
+	if followSymlink {
+		tarC = append(tarC, "-h")
+	}
+	tarC = append(tarC, "-", tarCArg)
```

Comment:
This creates `tar -c -f -h -` which is not the right syntax. This should be `tar  -h -c -f -` or something.

```console
# nerdctl --debug cp -L /tmp/hello-sym ubuntu-b1768:/hello-sym2-follow
nerdctl --debug cp -L /tmp/hello-sym ubuntu-b1768:/hello-sym2-follow
DEBU[0000] executing [/bin/tar -c -f -h - hello-sym] in "/tmp" 
DEBU[0000] executing [/bin/tar -x -f -] in "/proc/245919/root/hello-sym2-follow" 
/bin/tar: -: Cannot stat: No such file or directory
/bin/tar: Exiting with failure status due to previous errors
/bin/tar: This does not look like a tar archive
/bin/tar: Exiting with failure status due to previous errors
FATA[0000] failed to wait [/bin/tar -c -f -h - hello-sym]: exit status 2 
````

Reply:
fixed